### PR TITLE
feat: add push providers

### PR DESCRIPTION
**Pull request by peterdeme**
Add push providers

### CODE REVIEW

**Comment by peterdeme** [comment]
Agreed with Ferhat that we won't add tests until we don't generate test credentials and it can be merged as it is for now

**Review by yaziine — APPROVED** [review]
lgtm

File: app.go [hunk 1]
```diff
@@ -264,3 +264,56 @@ func (c *Client) RevokeTokens(ctx context.Context, before *time.Time) (*Response
...
+	DisabledAt     *time.Time `json:"disabled_at,omitempty"`
+	DisabledReason string     `json:"disabled_reason,omitempty"`
+
+	APNAuthKey              string `json:"apn_auth_key,omitempty"`
+	APNKeyID                string `json:"apn_key_id,omitempty"`
+	APNP12Cert              string `json:"apn_p12_cert,omitempty"`
+	APNTeamID               string `json:"apn_team_id,omitempty"`
+	APNTopic                string `json:"apn_topic,omitempty"`
+	APNAuthType             string `json:"apn_auth_type,omitempty"`
+	APNDevelopment          bool   `json:"apn_development,omitempty"`
```
Comment (driver-devel) [inline app.go:281]: We might want to not introduce this field if we are going to detect it automatically @ferhatelmas

Comment (ferhatelmas) [reply app.go:281]: yes, let's drop it

File: app.go [hunk 2]
```diff
@@ -264,3 +264,56 @@ func (c *Client) RevokeTokens(ctx context.Context, before *time.Time) (*Response
...
+type PushProvider struct {
+	Type           string     `json:"type"`
+	Name           string     `json:"name"`
+	Description    string     `json:"description,omitempty"`
+	DisabledAt     *time.Time `json:"disabled_at,omitempty"`
+	DisabledReason string     `json:"disabled_reason,omitempty"`
+
+	APNAuthKey              string `json:"apn_auth_key,omitempty"`
+	APNKeyID                string `json:"apn_key_id,omitempty"`
+	APNP12Cert              string `json:"apn_p12_cert,omitempty"`
```
Comment (ferhatelmas) [inline app.go:277]: let's drop this too

File: app.go [hunk 3]
```diff
@@ -264,3 +264,56 @@ func (c *Client) RevokeTokens(ctx context.Context, before *time.Time) (*Response
...
+	APNTeamID               string `json:"apn_team_id,omitempty"`
+	APNTopic                string `json:"apn_topic,omitempty"`
+	APNAuthType             string `json:"apn_auth_type,omitempty"`
```
Comment (ferhatelmas) [inline app.go:280]: this one too, backend will assume it

File: app.go [hunk 4]
```diff
@@ -264,3 +264,56 @@ func (c *Client) RevokeTokens(ctx context.Context, before *time.Time) (*Response
...
+	APNDevelopment          bool   `json:"apn_development,omitempty"`
+	APNHost                 string `json:"apn_host,omitempty"`
```
Comment (ferhatelmas) [inline app.go:282]: drop this, not supported

File: app.go [hunk 5]
```diff
@@ -264,3 +264,56 @@ func (c *Client) RevokeTokens(ctx context.Context, before *time.Time) (*Response
 	err := c.makeRequest(ctx, http.MethodPatch, "app", nil, setting, &resp)
 	return &resp, err
 }
+
+type PushProvider struct {
+	Type           string     `json:"type"`
```
Comment (ferhatelmas) [inline app.go:269]: let's define a type for this and provide constants

File: app.go [hunk 6]
```diff
@@ -264,3 +264,56 @@ func (c *Client) RevokeTokens(ctx context.Context, before *time.Time) (*Response
...
+
+	APNAuthKey              string `json:"apn_auth_key,omitempty"`
+	APNKeyID                string `json:"apn_key_id,omitempty"`
+	APNP12Cert              string `json:"apn_p12_cert,omitempty"`
+	APNTeamID               string `json:"apn_team_id,omitempty"`
+	APNTopic                string `json:"apn_topic,omitempty"`
+	APNAuthType             string `json:"apn_auth_type,omitempty"`
+	APNDevelopment          bool   `json:"apn_development,omitempty"`
+	APNHost                 string `json:"apn_host,omitempty"`
+	APNNotificationTemplate string `json:"apn_notification_template,omitempty"`
```
Comment (ferhatelmas) [inline app.go:283]: drop this since under v2, it's not there

File: app.go [hunk 7]
```diff
@@ -264,3 +264,56 @@ func (c *Client) RevokeTokens(ctx context.Context, before *time.Time) (*Response
...
+
+	FirebaseServerKey            string `json:"firebase_server_key,omitempty"`
+	FirebaseCredentials          string `json:"firebase_credentials,omitempty"`
+	FirebaseDataTemplate         string `json:"firebase_data_template,omitempty"`
```
Comment (ferhatelmas) [inline app.go:287]: drop this one too since not supported

File: app.go [hunk 8]
```diff
@@ -264,3 +264,56 @@ func (c *Client) RevokeTokens(ctx context.Context, before *time.Time) (*Response
...
+	FirebaseNotificationTemplate string `json:"firebase_notification_template,omitempty"`
```
Comment (ferhatelmas) [inline app.go:288]: we will use this one for clarity with v1, we will name differently so for this PR, drop it too

File: app.go [hunk 9]
```diff
@@ -264,3 +264,58 @@ func (c *Client) RevokeTokens(ctx context.Context, before *time.Time) (*Response
...
 	return &resp, err
 }
+
+type PushProviderType string
+
+const (
+	PushProviderTypeFirebase PushProviderType = "firebase"
+	PushProviderTypeAPN      PushProviderType = "apn"
+	PushProviderTypeHuawei   PushProviderType = "huawei"
+	PushProviderTypeXiaomi   PushProviderType = "xiaomi"
```
Comment (ferhatelmas) [inline app.go:274]: could you remove type probably?

File: app.go [hunk 10]
```diff
@@ -264,3 +264,58 @@ func (c *Client) RevokeTokens(ctx context.Context, before *time.Time) (*Response
...
+	Description    string           `json:"description,omitempty"`
+	DisabledAt     *time.Time       `json:"disabled_at,omitempty"`
+	DisabledReason string           `json:"disabled_reason,omitempty"`
+
+	APNAuthKey string `json:"apn_auth_key,omitempty"`
+	APNKeyID   string `json:"apn_key_id,omitempty"`
+	APNTeamID  string `json:"apn_team_id,omitempty"`
+	APNTopic   string `json:"apn_topic,omitempty"`
+
+	FirebaseServerKey   string `json:"firebase_server_key,omitempty"`
```
Comment (ferhatelmas) [inline app.go:289]: get rid of this too

File: app.go (hunk 9 above)
Comment (peterdeme) [reply app.go]: just realized this exists already in device.go ✅

**Review by ferhatelmas — APPROVED** [review]
LGTM

**Pull request merged**